### PR TITLE
Add option to pass block to controller's #save!

### PR DESCRIPTION
**Pull request by s-andringa**
Hi @pond, it's me [again](https://github.com/RIPAGlobal/scimitar/pull/71).

This time I extended `ActiveRecordBackedResourcesController#save!` to accept a block, so that subclasses may override the default `record.save!` call with their own saving logic, without having to reimplement the rescue clause for handling uniqueness errors. Also, it aligns nicely with the destroy action, which already accepts a block for a very similar reasons.

In our case we would like a service object to save the record and do some other stuff, because we do not want to employ AR callbacks for that other stuff. With these changes it can look like this:

```rb
def save!(record)
  super do
    Services::ManagedUser::Save.new(record).call
  end
end
```

This will be helpful in many situations. But sometimes it's not enough.

We are trying to make things work with an ActiveModel model that wraps an ActiveRecord record. It mostly delegates stuff to the wrapped record, but it has its own additional validations. That means that we'll need to rescue `ActiveModel::ValidationError` next to `ActiveRecord::InvalidRecord`. Therefore we need to override the rescue clause in #save!, but we'd still like to avoid copy-pasting all of the invalid record handling logic. Therefore I have extracted that logic to a separate method. This way we only need to copy-paste one method call, and we don't have to override the actual logic:

```rb
def save!(record)
  Services::ManagedUser::Save.new(record).call
rescue ActiveModel::ValidationError, ActiveRecord::InvalidRecord
  handle_invalid_record(record)
end
```

In short, both changes just add a bit more extensibility to the controller.

I would love to see this merged in one form or another. Thanks!

### CODE REVIEW

File: app/controllers/scimitar/active_record_backed_resources_controller.rb [hunk 1]
```diff
@@ -153,18 +153,31 @@ def record_to_scim(record)
...
+        if block_given?
+          yield(record)
+        else
+          record.save!
+        end
       rescue ActiveRecord::RecordInvalid => exception
+        handle_invalid_record(exception.record)
+      end
+
+      def handle_invalid_record(record)
```
Comment (pond) [inline app/controllers/scimitar/active_record_backed_resources_controller.rb:180]: This refactor is nice, but I'm of course going to ask for docs :joy: - just a brief RDoc comment, same pattern as `save!`  itself.

**Review by pond — CHANGES_REQUESTED** [review]
Looks great & only requesting changes on RDoc on the new method as I have to be a stickler for details like that in a public gem where docs are particularly important.

No other changes tho, on the money overall! Thanks.

**Comment by s-andringa** [comment]
Thanks @pond. I just pushed the RDoc comment for #handle_invalid_record.

Did you have a chance to look at my other PR https://github.com/RIPAGlobal/scimitar/pull/71 ?

**Comment by pond** [comment]
@s-andringa Yes, sorry it took a while. Only major request you'll see over there is a `README.md` addition.

Merging here, it's good to go. Thanks! I'll compile a few of the current PRs together and hope to do a new gem version release pretty soon.

**Review by pond — APPROVED** [review]

**Comment by s-andringa** [comment]
> I'll compile a few of the current PRs together and hope to do a new gem version release pretty soon.

That'd be great. Thank you.

**Comment by pond** [comment]
This is now available via RubyGems in v2.6.0 / v1.7.0.

**Comment by s-andringa** [comment]
Great, thanks @pond!
